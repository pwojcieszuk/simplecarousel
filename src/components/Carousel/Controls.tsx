import React from "react";
import styles from "components/Carousel/carousel.module.scss";
import { ControlsProps as Props } from "components/Carousel/logic/types";
import prepareButtons from "components/Carousel/logic/prepareButtons";

const Controls: React.FC<Props> = ({ dispatch, items, step }) => (
  <div className={styles.carouselBtns}>
    <button onClick={() => dispatch({ type: "goToPrevItem" })}>Prev</button>
    {prepareButtons(
      step,
      Object.keys(items).map((i) => parseInt(i))
    ).map((childIndex) => (
      <button
        key={childIndex}
        onClick={() => dispatch({ type: "goToItem", itemIndex: childIndex })}
        id={`${childIndex}`}
      />
    ))}
    <button onClick={() => dispatch({ type: "goToNextItem" })}>Next</button>
  </div>
);

export default Controls;
