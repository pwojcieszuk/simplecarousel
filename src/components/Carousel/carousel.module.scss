.carouselContainer {
  overflow: hidden;
}

.carouselSlider {
  height: 50%;
  max-height: 100vh;
}

.carouselItem {
  display: flex;
  width: 100%;
  justify-content: space-evenly;

  & > img {
    height: 200px;
    width: 200px;
  }
}

.carouselBtns {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;

  & > * {
    margin: 0 0.5rem;
  }
}
