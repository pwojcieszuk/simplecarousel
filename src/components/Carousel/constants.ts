import React from "react";
import { CarouselState, Action } from "components/Carousel/logic/types";
import goToPrevItem from "components/Carousel/state/goToPrevItem";
import goToNextItem from "components/Carousel/state/goToNextItem";
import autoPlay from "components/Carousel/state/autoPlay";
import goToItemAtIndex from "components/Carousel/state/goToItemAtIndex";

const cssTranslate = (value: number | string) => `translateX(${value})`;

export const DEFAULT_STYLE = (duration: number): React.CSSProperties => ({
  transition: `transform ${duration}ms ease-in-out`,
  transform: cssTranslate(0),
});

export const TRANSITION_STYLES_FORWARD = {
  entering: { transform: cssTranslate(0) },
  entered: { transform: cssTranslate("-100%") },
  exiting: { transform: cssTranslate("-100%") },
  exited: { transform: cssTranslate("-200%") },
} as { [key: string]: React.CSSProperties };

export const TRANSITION_STYLES_REVERSE = {
  entering: { transform: cssTranslate("-100%") },
  entered: { transform: cssTranslate(0) },
  exiting: { transform: cssTranslate(0) },
  exited: { transform: cssTranslate("100%") },
} as { [key: string]: React.CSSProperties };

export const ACTIONS_TYPES = {
  GO_TO_PREV_ITEM: "goToPrevItem",
  GO_TO_NEXT_ITEM: "goToNextItem",
  AUTO_PLAY: "autoPlay",
  GO_TO_ITEM_AT_INDEX: "goToItemAtIndex",
};

export const ACTIONS: {
  [key: string]: (state: CarouselState, action: Action) => CarouselState;
} = {
  goToPrevItem,
  goToNextItem,
  autoPlay,
  goToItemAtIndex,
};
