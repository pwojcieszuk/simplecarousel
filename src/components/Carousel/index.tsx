import React from "react";
// import styles from "components/Carousel/carousel.module.scss";

type Props = {
  children?: React.ReactChild | React.ReactChild[];
};

const Carousel: React.FC<Props> = ({ children }) => {
  return (
    <>
      {React.Children.map(
        children,
        (child) =>
          React.isValidElement(child) &&
          React.cloneElement(child, { className: "styles.carouselItem" })
      )}
    </>
  );
};

export default Carousel;
