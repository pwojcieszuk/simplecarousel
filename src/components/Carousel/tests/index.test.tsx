import React from "react";
import { shallow } from "enzyme";
import Carousel from "components/Carousel";

describe("Carousel", () => {
  it("Renders", () => {
    const wrapper = shallow(<Carousel>some child</Carousel>);
    expect(wrapper).toMatchSnapshot();
  });
});
