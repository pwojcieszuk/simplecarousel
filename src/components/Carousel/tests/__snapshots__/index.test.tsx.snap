// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel Renders 1`] = `<Fragment />`;
